src/WebHandlers.cpp

```diff
@@ -203,38 +207,45 @@ void AsyncStaticWebHandler::handleRequest(AsyncWebServerRequest* request) {
     // set etag to lastmod timestamp if available, otherwise to size
     String etag;
     if (lw) {
-      setLastModified(gmtime(&lw));
+      setLastModified(lw);
 #if defined(TARGET_RP2040)
       // time_t == long long int
-      const size_t len = 1 + 8 * sizeof(time_t);
+      constexpr size_t len = 1 + 8 * sizeof(time_t);
       char buf[len];
-      char* ret = lltoa(lw, buf, len, 10);
+      char* ret = lltoa(lw + request->_tempFile.size(), buf, len, 10);
       etag = ret ? String(ret) : String(request->_tempFile.size());
 #else
-      etag = String(lw);
+      etag = lw + request->_tempFile.size();   // etag combines file size and lastmod timestamp
 #endif
     } else {
-      etag = String(request->_tempFile.size());
+      etag = request->_tempFile.size();
     }
-    if (_last_modified.length() && _last_modified == request->header(T_IMS)) {
-      request->_tempFile.close();
-      request->send(304); // Not modified
-    } else if (_cache_control.length() && request->hasHeader(T_INM) && request->header(T_INM).equals(etag)) {
+
+    bool not_modified = false;
+
+    // if-not-modified has precedence over IMS
+    if (request->hasHeader(T_INM))
```

Comment:
I think it should be: `if (etag.length() && request->hasHeader(T_INM))`

> Unlike [If-Unmodified-Since](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Unmodified-Since), If-Modified-Since can only be used with a [GET](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET) or [HEAD](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD). When used in combination with [If-None-Match](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match), it is ignored, unless the server doesn't support If-None-Match.

Ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since

Reply:
`etag` is always set by this point, either to plan file size or to a mix of timestamp and size, so not needed to recheck

Reply:
> `etag` is always set by this point, either to plan file size or to a mix of timestamp and size, so not needed to recheck

No it won't be if `_tempFile.getLastWrite();` returns 0, which can be a possiiblity, then we don't fo in the `if (lw)`

Also, `response->addHeader(T_ETag, etag.c_str());` should only be added `if(etag.length())` otherwise a blank etag can be added, which does no comply wth the http spec (etag value is required)

Reply:
ehh....
are we looking at the same place?

```C++
    if (lw) {
       // bla-bla
    } else {
      etag = request->_tempFile.size();
    }
```
`etag` is either time^size or just size. So tag is always present. I'm not happy with "just size" but have no better ideas for now.

Reply:
I completely missed it! thanks!